templates/boltzmann/middleware/honeycomb.ts

```diff
@@ -9,77 +19,180 @@ import beeline from 'honeycomb-beeline'
-      const trace = beeline.startTrace({
-        [schema.EVENT_TYPE]: 'boltzmann',
-        [schema.PACKAGE_VERSION]: '1.0.0',
-        [schema.TRACE_SPAN_NAME]: `${context.method} ${context.url.pathname}${context.url.search}`,
-        [schema.TRACE_ID_SOURCE]: traceContext.source,
-        'request.host': context.host,
-        'request.original_url': context.url.href,
-        'request.remote_addr': context.remote,
-        'request.method': context.method,
-        'request.scheme': context.url.protocol,
-        'request.path': context.url.pathname,
-        'request.query': context.url.search
-      },
-      traceContext.traceId,
-      traceContext.parentSpanId,
-      traceContext.dataset)
-
-      if (isDev()) {
-        context._honeycombTrace = trace
-      }
+  if (isOtel(process.env)) {
+    return function honeycombOtelTrace (next: Handler) {
+      return async (context: Context) => {
+        const tracer = getOtelTracer()
+        let carrier = {}
 
-      if (traceContext.customContext) {
-        beeline.addContext(traceContext.customContext)
-      }
+        // Start a parent span
+        const parentSpan = tracer.startSpan(`${context.method} ${context.url.pathname}${context.url.search}`, {
+          attributes: {
+            [SemanticAttributes.HTTP_HOST]: context.host,
+            [SemanticAttributes.HTTP_URL]: context.url.href,
+            [SemanticAttributes.NET_PEER_IP]: context.remote,
+            [SemanticAttributes.HTTP_METHOD]: context.method,
+            [SemanticAttributes.HTTP_SCHEME]: context.url.protocol,
+            [SemanticAttributes.HTTP_ROUTE]: context.url.pathname,
+            [OTEL_REQ_QUERY]: context.url.search
+          }
+        })
 
-      if (!trace) {
-        return next(context)
-      }
+        // this propagator takes care of extracting trace parent
+        // and state from request headers (and so on)
+        const propagator = new W3CTraceContextPropagator()
 
-      const boundFinisher = beeline.bindFunctionToTrace((response: ServerResponse) => {
-        beeline.addContext({
-          'response.status_code': String(response.statusCode)
-        })
+        propagator.inject(
+          otelTrace.setSpanContext(
+            ROOT_CONTEXT,
+            parentSpan.spanContext()
+          ),
+          carrier,
+          defaultTextMapSetter
+        )
 
-        beeline.addContext({
-          'request.route': context.handler.route,
-          'request.method': context.handler.method,
-          'request.version': context.handler.version
-        })
+        // create a parent active context
+        const parentContext = propagator.extract(
+          ROOT_CONTEXT,
+          carrier,
+          defaultTextMapGetter
+        )
 
-        const params = Object.entries(context.params).map(([key, value]) => {
-          return [`request.param.${key}`, value]
+        // set the active context
+        otelContext.with(parentContext, () => {
```

Comment:
I was doing something slightly different here before, based on ci's otel.js. This is the API that's better documented in the otel docs, and assuming I have my concurrency straight I believe this will do the right thing.

Maybe I should think harder about how to unit test this? There's a `middleware/test.ts` file I'm not very familiar with - maybe stand up the middlewares in a stack and do some asserts on the attributes in test handlers/middleware?